### ISSUE TITLE
UX — a finding inside a shared hook is reported once per consuming component: 81% of corpus output is the same source line repeated

### ISSUE THREAD

**Issue by rboudrouss** [rboudrouss/reactant-analyzer#129]
Found while re-measuring the corpus for the 2026-09-02 diagnostic. This is the largest single distortion in the tool’s output and it is not recorded anywhere.

## The measurement

Findings vs **distinct source locations** (`file` + `line` + `col` + `message`),
one run of `check` per repository:

| corpus | reported | distinct | repeated |
|---|---|---|---|
| chakra-ui | 120 | 18 | 85% |
| mantine | 3165 | 359 | 89% |

The top of chakra-ui, grouped:

```
87 reports, 87 distinct components — use-chart.ts:123
12 reports, 12 distinct components — use-chart.ts:123   (a second message, same line)
 4 reports,  4 distinct components — use-media-query.ts:45
 1 report                          — color-picker-change-end.tsx:14
 …
```

**99 of 120 findings are one line of one shared hook.** The multiplicity is
exactly the number of components that inline it — 87 reports, 87 distinct
`component` values, one `file`/`line`.

## Why it happens

`program_result.components` is keyed by component name, so each component is
analysed once and the rule pass runs once per component — there is no
re-analysis and no double-counting bug. The multiplicity is honest: a custom
hook inlined into 87 components genuinely produces the finding 87 times, once
per component, each row carrying its own `component` / `component_file` and the
*hook’s* `file` / `line`.

What is missing is the collapse on the way out. Nothing in `driver` groups by
source location, so the human report prints the same line 87 times and the
summary counts it 87 times.

## Why it matters beyond noise

Every corpus number in the ADRs and issue history is a count of *consumer
attributions*, not of defects. "mantine: 3165 findings" reads as a defect
count and is off by roughly a factor of nine. Any future before/after
comparison inherits the distortion, and a regression that changes which
components inline a hook moves the number without changing a single defect.

## Fix direction

Group by `(rule, file, line, col, message)` in the human report and print the
consumer count — `use-chart.ts:123 … (in 87 components)` — with the component
list behind `--verbose` or `--trace`. Keep the JSON rows per component: a
consumer that wants per-component attribution has it today and should keep it.
The summary counts should follow whatever the human report shows, and
`--fail-on` should be unaffected either way.

Worth deciding at the same time whether the corpus measure scripts should quote
distinct locations from now on. I would say yes, and that the ADRs stay as
written with a note, rather than being retro-fitted.

**Comment by rboudrouss** [comment]
Full corpus table, `(rule, file, line, col, message)` as the key, one run per repository at `0c45de8`:

| corpus | reported | distinct | repeated |
|---|---|---|---|
| ai-chatbot | 10 | 9 | 10% |
| bulletproof-react | 3 | 3 | — |
| chakra-ui | 120 | 20 | 83% |
| commerce | 1 | 1 | — |
| dub | 1789 | 507 | 72% |
| excalidraw | 21 | 21 | — |
| mantine | 3165 | 359 | 89% |
| memos | 27 | 27 | — |
| next-shadcn-dashboard-starter | 21 | 19 | 10% |
| novel | 2 | 2 | — |
| precedent | 1 | 1 | — |
| shadcn-admin | 9 | 8 | 11% |
| twenty | 1056 | 386 | 63% |
| zustand | 2 | 2 | — |
| **total** | **6227** | **1365** | **78%** |

The repositories with no repetition are the ones with no widely-shared custom
hooks; the effect scales with how much a codebase factors its hooks, which is
to say it is worst on the codebases that are best factored.

**Comment by rboudrouss** [comment]
Title corrected: 81% is the corpus-wide figure with the full `(rule, file, line, col)` key. 89–90% is mantine specifically, which is what the original title took.

**Comment by rboudrouss** [comment]
Fixed in `b123be9`.

The human report now groups by `(rule, file, line, col, message)` — the key this issue measured with. Each distinct location prints once with `[in 87 components]`; `--trace` names the consumers; a component whose every finding was already printed is counted in a footer (`N component(s) hidden — every finding in them is a source line already reported above`) rather than repeated; and the summary counts locations, with the row total as a `— N component attribution(s)` tail so the collapse is visible rather than silent.

Grouping is a display fact and lives in the renderer, so the two things this issue asked to leave alone are untouched: **the JSON keeps one row per component** (schema v2 unchanged, `summary.warnings` still the row count) and **`--fail-on` reads the row counts**. A diagnostic with no range is never grouped — two positions we cannot name are not evidence of one location.

Measured:

| corpus | before | after |
|---|---:|---:|
| chakra-ui | 122 | 22 |
| mantine | 3165 | 359 |

`tests/location_grouping.rs` pins all five properties over `tests/fixtures/shared_hook_repeat` (one defect, one shared hook, three consumers); gated by removal — disabling the grouping turns exactly the three grouping tests red and leaves the JSON and `--fail-on` tests green.

On the last question: yes, the corpus measure scripts should quote distinct locations from now on, and the ADRs stay as written — `docs/limitations.md` carries the note.
